internal/tjson/schema.go

```diff
@@ -123,15 +122,66 @@ var (
 )
 
 // Equal returns true if the schemas are equal.
+// For composite types schemas are equal if their types and subschemas are equal.
+// For scalar types schemas are equal if their types and formats are equal.
 func (s *Schema) Equal(other *Schema) bool {
 	if s == other {
 		return true
 	}
 
-	// TODO compare significant fields only (ignore title, description, etc.)
-	// TODO compare format according to type (for example, for Number, EmptyFormat == Double)
-	// https://github.com/FerretDB/FerretDB/issues/683
-	return reflect.DeepEqual(s, other)
+	if s.Type != other.Type {
+		return false
+	}
+
+	// If `s` and `other` are objects, compare their properties.
+	if s.Type == Object {
+		if len(s.Properties) != len(other.Properties) {
+			return false
+		}
+		for k, v := range s.Properties {
+			vother, ok := other.Properties[k]
+			if !ok {
+				return false
+			}
+			if eq := v.Equal(vother); !eq {
+				return false
+			}
+		}
+		return true
+	}
+
+	// If `s` and `other` are arrays, compare their items.
+	if s.Type == Array {
+		if s.Items == nil || other.Items == nil {
+			panic("schema.Equal: array with nil items")
+		}
+		return s.Items.Equal(other.Items)
+	}
+
+	if s.Format != other.Format {
+		// Normalizing schemas: empty format is equal to double for numbers and int64 for integers,
+		// see https://docs.tigrisdata.com/overview/schema#data-types.
+		switch s.Type {
+		case Number:
+			if s.Format == EmptyFormat {
+				s.Format = Double
```

Comment:
We should not modify receiver in `Equal` method

Reply:
Ah, that answers the question of where data race is from! 🙈

Reply:
Fixed